File: ShadowFrameworkLite20/src/shadow/renderer/data/SFDataAsset.java
package shadow.renderer.data;

import java.util.List;

import shadow.system.SFException;
import shadow.system.SFInitiable;
import shadow.system.SFInitiator;
import shadow.system.data.SFDataObject;
import shadow.system.data.SFDataset;

/**
 * Abstract class for Assets which can be stored and retrieved through SFDataset
 * mechanisms. <br/>
 * Subclasses should be able to provide an implementation of the method required
 * to build up the resource from its data. <br/>
 * Each DataAsset will be built only once. <br/>
 * The Default {@link SFDataObject} is null and should be assigned through the
 * protected {@link setData} method.
 * 
 * @author Alessandro Martinelli
 * 
 * @param <T> the Type of graphical element which is kept by this asset.
 */
public abstract class SFDataAsset<T extends SFInitiable> extends SFAssetModule implements SFDataset, SFAsset<T> {

	private T resource = null;
	
	private static boolean updateMode=false;
	
	//The Number of active access to this resource
	private int resourceLock=0;
	
	private SFLibraryReference<?>[] references;
	
	/**
	 * Build up a valid resource using data content available from
	 * {@linkplain SFDataObject}. This method will be called only once
	 * 
	 * @return a valid resource of type T
	 */
	protected abstract T buildResource();

	@Override
	public T getResource() {
		if (resource == null || updateMode) {
			resource = buildResource();
			SFInitiator.addInitiable(resource);
		}
		resourceLock++;
		return resource;
	}
	
	public void releaseResource() throws SFException{
		if(resourceLock<=0 || resource==null)
			throw new SFException("A Resource has been released, but it was already unavailable");
		resourceLock--;
		if(resourceLock==0){
			if(references!=null){
				for (int i = 0; i < references.length; i++) {
					SFDataAsset<?> asset=(references[i].getDataset());
					asset.releaseResource();
				}
			}
			SFInitiator.addDestroyable(resource);
			resource=null;	
		}
	}
	
	@SuppressWarnings("all")
	protected void setReferences(SFLibraryReference... references){
		this.references=references;
	}
	
	@SuppressWarnings("all")
	public void updateReferences(List<SFLibraryReference> referencesList) {
		SFLibraryReference[] references=new SFLibraryReference[referencesList.size()];
		for (int i = 0; i < references.length; i++) {
			references[i]=referencesList.get(i);
		}
		setReferences(references);
	}
	

	public static boolean isUpdateMode() {
		return updateMode;
	}

	public static void setUpdateMode(boolean updateMode) {
		SFDataAsset.updateMode = updateMode;
	}
	
}

File: ShadowFrameworkJS/src/shadow/renderer/data/SFDataAsset.js

var SFDataAsset_updateMode=false;

function SFDataAsset(){
	this.resource=null;
	this.resourceLock=0;
	this.references=new Array();
}

inherit(SFDataAsset,SFAssetModule);


SFDataAsset.prototype["getResource"]=function(){
			if (this.resource == null || SFDataAsset_updateMode) {
				this.resource = this.buildResource();
				SFInitiator_addInitiable(this.resource);
			}
			this.resourceLock++;
			return this.resource;
		};


SFDataAsset.prototype["releaseResource"]=function(){
			if(this.resourceLock<=0 || this.resource==null)
				throw "A Resource has been released, but it was already unavailable";
			this.resourceLock--;
			if(this.resourceLock==0){
				if(this.references!=null){
					for (var i = 0; i < this.references.size(); i++) {
						var asset=(references[i].getDataset());
						asset.releaseResource();
					}
				}
				SFInitiator.addDestroyable(resource);
				resource=null;	
			}
		};


SFDataAsset.prototype["setReferences"]=function(references){
			this.references=references;
		};

SFDataAsset.prototype["updateReferences"]=function(referencesList){
			var references=new Array();
			for (var i = 0; i < referencesList.length; i++) {
				references[i]=referencesList.get(i);
			}
			this.setReferences(references);
		};
		
function SFDataAsset_isUpdateMode(){
	return SFDataAsset_updateMode;
}

function SFDataAsset_setUpdateMode(updateMode){
	SFDataAsset_updateMode=updateMode;
}

